Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -318,4 +319,55 @@ theorem lt_lxor'_cases {a b c : ℕ} (h : a < lxor' b c) : lxor' a c < b ∨ lxo
   (or_iff_right fun h' => (h.asymm h').elim).1 <| lxor'_trichotomy h.ne
 #align nat.lt_lxor_cases Nat.lt_lxor'_cases
 
+lemma div2_add_two (x) :
+    div2 (x + 2) = div2 x + 1 := by
+  simp only [div2_succ, bodd_succ, Bool.cond_not]
+  cases bodd x <;> simp only [cond_false, cond_true]
+
+lemma div2_eq_zero {x : ℕ} :
+    div2 x = 0 → x = 0 ∨ x = 1 := by
+  intro h
+  rcases x with ⟨⟩ | ⟨⟩ | x
+  · apply Or.inl rfl
+  · apply Or.inr rfl
+  · rw[div2_add_two] at h
+    contradiction
+
+lemma bitwise'_bit' {f : Bool → Bool → Bool} (a : Bool) (m : Nat) (b : Bool) (n : Nat)
+    (ham : m = 0 → a = true) (hbn : n = 0 → b = true) :
+    bitwise' f (bit a m) (bit b n) = bit (f a b) (bitwise' f m n) := by
+  rw [bitwise', binaryRec_eq', binaryRec_eq']
+  · apply Or.inr hbn
+  · apply Or.inr ham
+
+lemma bitwise_eq_bitwise' (f) : bitwise f = bitwise' f := by
+  funext x y
+  induction' hk : x + y using Nat.strongInductionOn with k ih generalizing x y
+  by_cases h1 : x = 0 <;> by_cases h2 : y = 0
+  · unfold bitwise
```

Comment:
You shouldn't be unfolding `bitwise` here: if you need to unfold it, that means a simpler lemma (where `unfold` is fine) is missing.

Reply:
Which is to say: we should probably have the lemmas:
* `bitwise_zero_left : bitwise f 0 m = if f false true then m else 0`
* `bitwise_zero_right : bitwise f n 0 = if f true false then n else 0
* `bitwise_of_ne_zero (n \ne 0) (m \ne 0) : ...`

Reply:
I've cleaned up the proof. Turns out we also needed `bitwise'_zero_right : bitwise' f m 0 = if f true false then m else 0`.
This result exists in `Mathlib.Init.Data.Nat.Bitwise`, but it has `f false false = false` as an assumption.
I've added an alternative theorem with a proof that does not need this assumption, but instead relies on `eq_mpr_eq_cast` from `Mathlib.Logic.Basic`.

This duplication seems undesirable, but my understanding is that `Init` files sould not depend on non-`Init` files.
Then again, do we even need this result in `Init.Data.Nat.Bitwise`? I would guess it's a historic artifact from lean3 where these definitions were the main way to interact with `Nat`s.

Reply:
I think most of the `Mathlib.Init` files are vestigial from lean3, and moving things out of them is totally fine